newlib: define _POSIX_TIMERS to 0 for micro-ROS

hippo5329 opened the pull request:
micro-ROS uses clock_gettime() and added a custom one when there is no real posix timers support. The latest newlib missed the func proto and caused build error. This patch fixed this issue. Tested on both pico and pico2.

In this timer.h
    #if defined(_POSIX_TIMERS)
    ...
    int clock_settime (clockid_t clock_id, const struct timespec *tp);
    int clock_gettime (clockid_t clock_id, struct timespec *tp);

hippo5329 committed:
newlib: define _POSIX_TIMERS to 0 for micro-ROS

micro-ROS uses clock_gettime() and added a custom one when there is no
real posix timers support. The latest newlib missed the func proto and
caused build error. This patch fixed this issue. Tested on both pico
and pico2.

In include/timer.h
    #if defined(_POSIX_TIMERS)
    ...
    int clock_settime (clockid_t clock_id, const struct timespec *tp);
    int clock_gettime (clockid_t clock_id, struct timespec *tp);

Signed-off-by: Thomas Chou <thomas@wytron.com.tw>

earlephilhower reviewed (changes requested):
Why not just set the define as part of your build process?  I'm not going to be rebuilding the toolchain for quite some time.

This would be better in `sys/features.h` where these kinds of flags are set.

Also I'd set it to `1` (or some POSIX stamp) not `0` because the checks are only an `if defined(_POSIX_TIMERS)` and don't look at the value.  At first glance here it seems like you're trying to *exclude* the functions by setting a flag to 0 but  in reality setting the flag to 0 enables them...

earlephilhower commented:
Alternatively, you can set the `#define` in your specific source code before including `time.h` and it will pull in the definitions without changing any build process or tools...

````
#define _POSIX_TIMERS 1
#include <time.h>
#include <...other stuff>
...
````

hippo5329 commented:
I tried the this approach and others before, but they did not work. Because ROS uses colcon to build packages. This _POSIX_TIMERS flag is ignored. The ROS project is large. We are not able to update them for pico only. Esp32 does not have this issue, because it supports posix timers. So far, we see this issue only on pico/pico2. This issue happened since Sep. There was no issue before.

hippo5329 committed:
gmp version 6.2.1

Because gmp 6.1.2 is removed from GNU repo, it failed to download.
Now it is 6.2.1.

Signed-off-by: Thomas Chou <thomas@wytron.com.tw>

earlephilhower commented on Makefile:
```diff
@@ -287,7 +287,7 @@ install = $(call arena,$(1))/$(ARCH)
 
 # GNU infra
-GMP_VER := 6.1.2
+GMP_VER := 6.2.1
```
Was this intentional?  Why?

hippo5329 replied:
Because gmp 6.1.2 is removed from GNU repo. It failed to download. Now it is 6.2.1.

earlephilhower commented on patches/lib-posix-timers.patch:
```diff
@@ -0,0 +1,16 @@
+ 
++#ifndef _POSIX_TIMERS
++#define _POSIX_TIMERS 0
++#endif
```
This define belongs in sys/features.h for consistency.  That's where flags like this are set for the whole library.

hippo5329 replied:
yes. I tried sys/features.h before. Both work. I will revert to this. Thanks.

earlephilhower commented:
Ah, I guess adding the define to the compile command line may be an issue with some build tools.  Cmake, it's easy.

Anyway, you can definitely define the value in your source file before the `#include <time.h>` and it will work.  If this is only in one or two files, that will get you going immediately.  If every source file separately includes time.h, yeah that would be a pain!

earlephilhower reviewed (approved):
I'm OK with this change now, thanks for working with me on it.

I may change the `#define` to a `1` just for sanity's sake (it's only used as an `ifdef` and actual value ignored, but eyeballing this in a couple years I'm sure I'd think this was *disabling* those calls, not actually *enabling* them).

hippo5329 commented:
Thanks a lot.

This is a special case to define it as 0. That means posix timer is NOT supported according to the documents. We only need the func proto to add a custom clock_gettime(). Yes, time.h is widely used.
